File: frontend/src/finance/api/reports.ts
/**
 * Reports API — campaign ROI, customer lifecycle, department performance.
 */
import client from "./client";

// ── Types ──

export interface CampaignROIItem {
  campaign_name: string;
  calls_generated: number;
  contacts_created: number;
  customers_converted: number;
  revenue_attributed: number;
  conversion_rate: number;
  avg_deal_size: number;
}

export interface CustomerLifecycleItem {
  contact_id: string;
  contact_name: string | null;
  contact_phone: string | null;
  contact_email: string | null;
  first_call_date: string | null;
  first_invoice_date: string | null;
  is_new_customer: boolean;
  lifetime_revenue: number;
  status: string | null;
}

export interface DepartmentPerformanceItem {
  department: string;
  calls_handled: number;
  contacts_generated: number;
  revenue_attributed: number;
  avg_duration_s: number;
}

// ── API calls ──

export async function getCampaignROI(
  businessId: string,
  days: number = 30,
): Promise<CampaignROIItem[]> {
  const res = await client.get("/reports/campaign-roi", {
    params: { business_id: businessId, days },
  });
  return res.data;
}

export async function getCustomerLifecycle(
  businessId: string,
): Promise<CustomerLifecycleItem[]> {
  const res = await client.get("/reports/customer-lifecycle", {
    params: { business_id: businessId },
  });
  return res.data;
}

export async function getDepartmentPerformance(
  businessId: string,
  days: number = 30,
): Promise<DepartmentPerformanceItem[]> {
  const res = await client.get("/reports/department-performance", {
    params: { business_id: businessId, days },
  });
  return res.data;
}

// ── Pipeline Funnel ──

export interface FunnelStage {
  stage: string;
  label: string;
  total: number;
  new_customers: number;
  returning_customers: number;
  from_campaigns: number;
  manual: number;
  revenue: number;
  conversion_pct: number;
}

export interface CampaignAttribution {
  campaign_name: string;
  channel: string | null;
  calls: number;
  leads: number;
  jobs: number;
  revenue: number;
  new_customers: number;
  returning_customers: number;
}

export interface PipelineFunnelResponse {
  stages: FunnelStage[];
  campaigns: CampaignAttribution[];
  totals: Record<string, number>;
  period_days: number;
}

export async function getPipelineFunnel(
  businessId: string,
  days: number = 30,
): Promise<PipelineFunnelResponse> {
  const res = await client.get("/reports/pipeline-funnel", {
    params: { business_id: businessId, days },
  });
  return res.data;
}

File: app/finance/schemas/reports.py
"""Pydantic schemas for Reports tab: campaign ROI, customer lifecycle, department performance."""

from uuid import UUID
from typing import Optional
from datetime import datetime

from pydantic import BaseModel


class CampaignROIItem(BaseModel):
    """One campaign's ROI metrics."""
    campaign_name: str
    calls_generated: int = 0
    contacts_created: int = 0
    customers_converted: int = 0
    revenue_attributed: float = 0
    conversion_rate: float = 0
    avg_deal_size: float = 0


class CustomerLifecycleItem(BaseModel):
    """One contact's lifecycle analysis: first call vs. first invoice."""
    contact_id: UUID
    contact_name: Optional[str] = None
    contact_phone: Optional[str] = None
    contact_email: Optional[str] = None
    first_call_date: Optional[datetime] = None
    first_invoice_date: Optional[datetime] = None
    is_new_customer: bool = True  # True if first invoice >= first call
    lifetime_revenue: float = 0
    status: Optional[str] = None  # prospect | active_customer | churned


class DepartmentPerformanceItem(BaseModel):
    """One department's performance metrics."""
    department: str
    calls_handled: int = 0
    contacts_generated: int = 0
    revenue_attributed: float = 0
    avg_duration_s: float = 0


# ── Pipeline Funnel ──


class FunnelStage(BaseModel):
    """Metrics for one pipeline stage."""
    stage: str  # calls | leads | jobs_created | jobs_completed | revenue
    label: str  # Human-readable label
    total: int = 0
    new_customers: int = 0
    returning_customers: int = 0
    from_campaigns: int = 0  # Attributed to a tracking number / campaign
    manual: int = 0  # No campaign attribution (manual entry, walk-in, etc.)
    revenue: float = 0  # Revenue attributed to contacts at this stage
    conversion_pct: float = 0  # % of previous stage that reached this stage


class CampaignAttribution(BaseModel):
    """Per-campaign pipeline attribution."""
    campaign_name: str
    channel: Optional[str] = None
    calls: int = 0
    leads: int = 0
    jobs: int = 0
    revenue: float = 0
    new_customers: int = 0
    returning_customers: int = 0


class PipelineFunnelResponse(BaseModel):
    """Complete pipeline funnel with attribution."""
    stages: list[FunnelStage]
    campaigns: list[CampaignAttribution]
    totals: dict[str, float]  # total_calls, total_leads, total_revenue, new_revenue, returning_revenue
    period_days: int
